lib/Runtime/Base/ThreadContext.cpp

```diff
@@ -3230,6 +3230,53 @@ ThreadContext::InvalidateProtoInlineCaches(Js::PropertyId propertyId)
     }
 }
 
+void
+ThreadContext::InvalidateMissingPropertyInlineCaches(const Js::Type *type, Js::PropertyId propertyId) {
+    InlineCacheList *inlineCacheList;
+    if (protoInlineCacheByPropId.TryGetValue(propertyId, &inlineCacheList))
+    {
+        if (PHASE_TRACE1(Js::TraceInlineCacheInvalidationPhase))
+        {
+            Output::Print(_u("InlineCacheInvalidation: invalidating missing-property proto caches for property %s(%u), type %d\n"),
+                GetPropertyName(propertyId)->GetBuffer(), propertyId, type->GetTypeId());
+            Output::Flush();
+        }
+
+        // Next section is based on InvalidateAndDeleteInlineCacheList.  Unlike that function, however, we only invalidate and
+        // remove entries that describe missing properties.
+        uint numCachesInvalidated = 0;
+        FOREACH_SLISTBASE_ENTRY_EDITING(Js::InlineCache*, inlineCache, inlineCacheList, editingIterator)
+        {
+            // ThreadContext::InvalidateAndDeleteInlineCacheList, on which this method is based, just deletes the entire list, so
+            // (among other things) it must update this->unregisteredInlineCacheCount to reflect the number of
+            // NULL entries in the list that are being deleted.  For simplicity, I skip those here and leave the count intact.
+            if (inlineCache != nullptr && inlineCache->u.proto.isProto && inlineCache->u.proto.type == type) {
```

Comment:
Should I remove entries where inlineCache is nullptr?  That seems like an unrelated cleanup operation best left to another method, but I'm open to arguments to the contrary.

Reply:
Have you seen null pointers in the list? I'd be curious to know where they're coming from.

Reply:
It would also be interesting to know if there are pointers to inline caches in the proto list that are not actually proto inline caches.

Reply:
I haven't specifically seen either null ptrs or non-proto inline caches in the proto list, but I haven't really looked.  I just basically copied this code from InvalidateAndDeleteInlineCacheList, which does similar checks.  When I get a chance, though, I'll instrument the code and take a look.